zkvm/src/types.rs

```diff
@@ -300,12 +342,61 @@ impl Value {
 }
 
 impl Expression {
-    pub fn constant<S: Into<Scalar>>(a: S) -> Self {
-        let a: Scalar = a.into();
+    pub fn constant<S: Into<ScalarWitness>>(a: S) -> Self {
+        Expression::Constant(a.into())
+    }
+}
+
+impl Neg for Expression {
+    type Output = Expression;
+
+    fn neg(self) -> Expression {
+        match self {
+            Expression::Constant(a) => Expression::Constant(-a),
+            Expression::Terms(mut terms, assignment) => {
+                for (_, n) in terms.iter_mut() {
+                    *n = -*n;
+                }
+
+                let x = match assignment {
+                    Some(a) => Some(-a),
+                    None => None,
+                };
+                Expression::Terms(terms, x)
+            }
+        }
+    }
+}
 
-        Expression {
-            terms: vec![(r1cs::Variable::One(), a)],
-            assignment: Some(ScalarWitness::Scalar(a)),
+impl Add for Expression {
+    type Output = Expression;
+
+    fn add(self, rhs: Expression) -> Expression {
+        match (self, rhs) {
+            (Expression::Constant(left), Expression::Constant(right)) => {
+                Expression::Constant(left + right)
+            }
+            (Expression::Constant(left), Expression::Terms(mut right_terms, right_assignment)) => {
+                // prepend constant term to `term vector` in non-constant expression
+                right_terms.insert(0, (r1cs::Variable::One(), left.into()));
```

Comment:
We double-checked that the order of terms in the linear combination does not affect the proof, only the order of constraints themselves and order of allocated variables does. So you can do `push()` instead of `insert()` here.

Reply:
However, it's probably cleaner+safer to keep it as is — this way the result is locally stable, which is easier to reason about.